staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/README.md

```diff
@@ -0,0 +1,68 @@
+# Azure Active Directory plugin for client authentication
+
+This plugin provides an integration with [Azure CLI](https://github.com/Azure/azure-cli) and Azure Active Directory device flow.
+
+## Usage
+
+### Microsoft Azure CLI 2.0
+
+1. Configure the `apiserver` to use the Azure Active Directory as an OIDC provider with following options
+
+   ```
+   --oidc-client-id="https://management.core.windows.net/" \
+   --oidc-issuer-url="https://sts.windows.net/TENANT_ID/"
+   --oidc-username-claim="sub"
+   ```
+
+   * Replace `TENANT_ID` with your tenant ID.
+
+2. Login with Azure CLI
+
+  ```
+  az login
+  ```
+
+3. Configure the `kubectl` to use the `azure` authentication provider
+
+   ```
+   kubectl config set-credentials "USER_NAME" --auth-provider=azure
+   ```
+
+   * Replace `USER_NAME` with your user name.
+
+### Device Flow
+
+1. Create an Azure native application following these [instructions](https://docs.microsoft.com/en-us/azure/active-directory/active-directory-app-registration)
+
+  Assign permissions to this application to access the `https://management.core.windows.net/"` audience.
+
+2. Configure the `apiserver` to use the Azure Active Directory as an OIDC provider with following options
+
+   ```
+   --oidc-client-id="https://management.core.windows.net/" \
```

Comment:
Let's take an example. I have a resource application registered with AAD for apiserver which has the resourceAppId `fdfb78c2-788c-4ab6-bbed-cfd89cd21025`. This is will be provided in the `--oidc-client-id`.  Now if I need to acquire a token for this application, another client application must be registered  wiht appId `319e3b43-c214-43e8-8d8d-8d33b8e47902` which has delegated permissions to the apiserver application. 

I can request a token with the following values:
-`applicatoinId`: `319e3b43-c214-43e8-8d8d-8d33b8e47902`
- `tenantId`: 
- `resource`:  `fdfb78c2-788c-4ab6-bbed-cfd89cd21025`

The returned token will have these claims (many claims ignored for brevity):

```
{
"aud": "spn:fdfb78c2-788c-4ab6-bbed-cfd89cd21025",
 "iss": "https://sts.windows.net/26f087ba-1a21-11e7-93ae-92361f002671/",
 "appid": "319e3b43-c214-43e8-8d8d-8d33b8e47902"
}
```

azure cli does the same but the audience is an URL: 

```
{
"aud": "https://management.core.windows.net/",
 "iss": "https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/",
  "appid": "04b07795-8ddb-461a-bbee-02f9e1bf7b46"
}
```
As you see from the first example the audience can be almost the same like the client id if one will use the UUID of the resource when acquiring the token. 

Note that any owner of an application can set `delegated permissions` to any `public resource application` registered in the same AAD.  The conclusion is that the RBAC policy should be based on the user and group claims. The audience does not have strong security guarantees.